src/Entities/Player.cpp

```diff
@@ -1231,6 +1218,30 @@ void cPlayer::SetGameMode(eGameMode a_GameMode)
 void cPlayer::LoginSetGameMode( eGameMode a_GameMode)
 {
 	m_GameMode = a_GameMode;
+
+	SetCapabilities();
+}
+
+
+
+
+
+void cPlayer::SetCapabilities()
+{
+	if (!IsGameModeCreative() || IsGameModeSpectator())
+	{
+		SetFlying(false);
+		SetCanFly(false);
+	}
+
+	if (!IsGameModeSpectator() && !IsVisible())
```

Comment:
Probably, I'm just not sure of the side-effects of calling SetVisible and calling it again for a player who's already in the state they are being set to may have unintended consequences (performance, client bugs, etc).